shpc/main/modules/base.py

```diff
@@ -410,3 +415,44 @@ def install(self, name, tag=None, **kwargs):
             name = "%s:%s" % (name, tag.name)
         logger.info("Module %s was created." % name)
         return container_path
+
+    # Module software can choose how to handle each of these cases
+    def _no_default_version(self, version_file, tag):
+        return
+
+    def _module_sys_default_version(self, version_file, tag):
+        return
+
+    def _set_default_version(self, version_file, tag):
+        """
+        Set the default version to the given tag
+        """
+        template = self._load_template("default_version")
+        utils.write_file(version_file, template.render(version=tag))
+
+    def write_version_file(self, uri, latest_tag_installed=None):
+        """
+        Create the .version file, if there is a template for it.
+        """
+        version_dir = os.path.join(self.settings.module_base, uri)
+        version_file = os.path.join(version_dir, ".version")
+
+        # No default versions
+        if self.settings.default_version in [False, None]:
+            return self._no_default_version(version_file, latest_tag_installed)
+
+        # allow the module software to control versions
+        if self.settings.default_version in [True, "module_sys"]:
+            return self._module_sys_default_version(version_file, latest_tag_installed)
+
+        # First or last installed
+        # The versions we actually have
+        found = [x for x in os.listdir(version_dir) if x != ".version"]
+        if len(found) == 1:
+            tag = found[0]
+        else:
+            selector = min if self.settings.default_version == "first_installed" else max
+            tag = selector(found, key=lambda x: utils.creation_date(os.path.join(version_dir, x)))
```

Comment:
There is a potential optimisation here: if `self.settings.default_version == "last_installed"` and `latest_tag_installed` is given, we can directly use it, bypassing the need to list the directory content and fetch their creation date

Reply:
I almost mentioned this in my review! I'm happy if you want to add it, but okay if not.